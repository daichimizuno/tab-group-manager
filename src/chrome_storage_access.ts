export default class ChromeStorageAccess {

  getAllTabGroup = (): Promise<string[]> => {
    return new Promise((resolve, reject) => {
      chrome.storage.sync.get(["tabGroup"], (value) => {
        if (chrome.runtime.lastError) {
          reject(chrome.runtime.lastError);
        }

        resolve(value.tabGroup as string[]);
      });
    });
  };

  addNewGroup = async (tabName: string) => {
    const tabGroupData = (await this.getAllTabGroup()) as string[] || [];
    console.log(`tab group data: ${JSON.stringify(tabGroupData)}`);

    // 既存のタブグループに新しいタブグループを追加
    tabGroupData.push(tabName)

    chrome.storage.sync.set({ "tabGroup": tabGroupData }, async () => {
      console.log(`new tab group is set, name:${tabName}`);
    });
  };
}
