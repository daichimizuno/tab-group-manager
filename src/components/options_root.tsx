import React from "react";
import ReactDOM from "react-dom";
import Options from "./option_page";

ReactDOM.render(
  <React.StrictMode>
    <Options />
  </React.StrictMode>,
  document.getElementById("root")
);
