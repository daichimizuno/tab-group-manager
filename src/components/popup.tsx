import React, { useEffect, useState } from "react";
import ReactDOM from "react-dom";

const Popup = () => {
  console.log("popup is opened");

  const [count, setCount] = useState(0);
  const [currentURL, setCurrentURL] = useState<string>();

  useEffect(() => {
    chrome.action.setBadgeText({ text: count.toString() });
  }, [count]);

  useEffect(() => {
    chrome.tabs.query({ active: true, currentWindow: true }, function (tabs) {
      setCurrentURL(tabs[0].url);
    });
  }, []);

  const changeBackground = () => {
    console.log("#changeBackground");
    chrome.tabs.query({ active: true, currentWindow: true }, function (tabs) {
      const tab = tabs[0];
      if (tab.id) {
        chrome.tabs.sendMessage(
          tab.id,
          {
            color: "#555555",
          },
          (msg) => {
            console.log("result message:", msg);
          }
        );
      }
    });
  };

  const createTab = () => {
    chrome.tabs.create({
      url: "https://readouble.com/laravel/5.4/ja/mix.html",
    });
  };

  const openOptionPage = () => {
    chrome.runtime.openOptionsPage(() => {
      console.log("option page is created");
    });
  };

  return (
    <>
      <button onClick={openOptionPage}>option page</button>

      <ul style={{ minWidth: "500px" }}>
        <li>Current URL: {currentURL}</li>
        <li>Current Time: {new Date().toLocaleTimeString()}</li>
      </ul>
      <button
        onClick={() => setCount(count + 1)}
        style={{ marginRight: "5px" }}
      >
        count up
      </button>
      <button onClick={createTab}>change background</button>
    </>
  );
};

ReactDOM.render(
  <React.StrictMode>
    <Popup />
  </React.StrictMode>,
  document.getElementById("root")
);
