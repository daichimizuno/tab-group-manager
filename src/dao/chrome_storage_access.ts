import { assertIsDefined, AssertIsDefinedError } from "../utils/validations";
import { StorageKeys } from "./storage_keys";

export interface TabGroup {
  id: number;
  tabGroupName: string;
  tabColor: Color;
  urls: string[];
}

const Color = {
  Grey: "grey",
  Blue: "blue",
  Red: "red",
  Yellow: "yellow",
  Green: "green",
  Pink: "pink",
  Purple: "purple",
  Cyan: "cyan",
} as const;
export type Color = typeof Color[keyof typeof Color];

export default class ChromeStorageAccess {
  addNewTabGroup = async (
    tabName: string,
    tabColor: Color
  ): Promise<Boolean> => {
    const tabGroupData = ((await this.getAllTabGroup()) as TabGroup[]) || [];

    // 既に他に同じ名前のタブグループ名がある場合は追加しない
    const hasTheSameGroupNameData =
      tabGroupData.findIndex(
        (tabGroup) => tabGroup.tabGroupName === tabName
      ) !== -1;
    if (hasTheSameGroupNameData) {
      return Promise.resolve(false);
    } else {
      const tabGroupLastIndex =
        ((await this.syncUpdateLastIndex()) as number) || 1;
      const newTabGroup = {
        id: tabGroupLastIndex,
        tabGroupName: tabName,
        tabColor: tabColor,
        urls: [],
      } as TabGroup;

      tabGroupData.push(newTabGroup);

      await chrome.storage.sync.set({ tabGroup: tabGroupData }, async () => {});

      return Promise.resolve(true);
    }
  };

  addUrlToTabGroup = async (url: string, tabId: number) => {
    const tabGroupData = ((await this.getAllTabGroup()) as TabGroup[]) || [];
    if (tabGroupData.length > 0) {
      const tabGroup = tabGroupData.find((tabGroup) => tabGroup.id === tabId);
      const tabGroupIndex = tabGroupData.findIndex(
        (tabGroup) => tabGroup.id === tabId
      );

      try {
        assertIsDefined(tabGroup);

        // 既存のTabGroupにurlを追加する
        tabGroup.urls.push(url);

        // TabGroupのリストを新しいものに入れ替える
        tabGroupData.splice(tabGroupIndex, 1, tabGroup);

        await chrome.storage.sync.set(
          { tabGroup: tabGroupData },
          async () => {}
        );
      } catch (e) {
        if (e instanceof AssertIsDefinedError) {
          throw new AssertIsDefinedError("不明なエラー");
        }
      }
    }
  };

  deleteTabGroup = async (tabNames: string[]) => {
    const tabGroupData = ((await this.getAllTabGroup()) as TabGroup[]) || [];

    tabNames.forEach((tabName) => {
      const targetTabGroupIndex = tabGroupData.findIndex(
        (tabGroup) => tabGroup.tabGroupName === tabName
      );

      if (targetTabGroupIndex !== -1) {
        tabGroupData.splice(targetTabGroupIndex, 1);
      }
    });

    await chrome.storage.sync.set({ tabGroup: tabGroupData }, async () => {});
  };

  syncUpdateLastIndex = async (): Promise<number> => {
    let index = await this.getTabGroupLastIndex();
    if (isNaN(index)) {
      index = 1;
    } else {
      index = index + 1;
    }

    await this.setTabGroupLastIndex(index.toString());
    return index;
  };

  getTabGroupLastIndex = (): Promise<number> => {
    return new Promise((resolve, reject) => {
      chrome.storage.sync.get([StorageKeys.TAB_GORUP_LAST_INDEX], (value) => {
        if (chrome.runtime.lastError) {
          throw chrome.runtime.lastError;
        }

        const index =
          value.tabGroupLastIndex === undefined
            ? NaN
            : Number(value.tabGroupLastIndex);
        resolve(index);
      });
    });
  };

  setTabGroupLastIndex = (newIndex: string): Promise<void> => {
    return new Promise((resolve, reject) => {
      chrome.storage.sync.set({ tabGroupLastIndex: newIndex }, () => {
        if (chrome.runtime.lastError) {
          throw chrome.runtime.lastError;
        }

        resolve();
      });
    });
  };

  getAllTabGroup = (): Promise<TabGroup[]> => {
    return new Promise((resolve, reject) => {
      chrome.storage.sync.get([StorageKeys.TAB_GROUP], (value) => {
        if (chrome.runtime.lastError) {
          reject(chrome.runtime.lastError);
        }

        resolve(value.tabGroup as TabGroup[]);
      });
    });
  };
}
