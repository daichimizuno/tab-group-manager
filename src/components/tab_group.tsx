import React from "react";
import { TabGroup } from "../chrome_storage_access";

export const TabGroupComponent = (props: any) => {
  const tabGroup = props.tabGroup;

  return (
    <>
      <table>
        <thead className="table-dark">
          <tr>
            <th scope="col-1">id</th>
            <th scope="col-2">色</th>
            <th scope="col-3">タブ名</th>
            <th scope="col-4">url</th>
          </tr>
        </thead>
        <tbody>
          {tabGroup.map((tab: TabGroup, index: number) => {
            return (
              <tr key={index}>
                <td data-testid={tab.id}>{tab.id}</td>
                <td data-testid={tab.id}>{tab.tabColor}</td>
                <td data-testid={tab.id}>{tab.tabGroupName}</td>
                <td data-testid={tab.id}>{tab.urls.map((url) => url)}</td>
              </tr>
            );
          })}
        </tbody>
      </table>
    </>
  );
};
