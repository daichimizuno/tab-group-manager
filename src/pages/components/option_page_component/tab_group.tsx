import { Typography } from "@mui/material";
import React from "react";
import { TabGroup } from "../../../dao/chrome_storage_access";
import TabGroupTable from "./tab_group_table";

const TabGroupComponent = (props: any) => {
  const tabGroup = props.tabGroup as TabGroup[];
  const deleteTabGroup = props.deleteTabGroup;

  return (
    <>
      <Typography data-testid="title">
        <h3>登録されたタブグループ</h3>
      </Typography>

      <TabGroupTable
        tabGroup={tabGroup}
        deleteTabGroup={(tabNames: string[]) => deleteTabGroup(tabNames)}
      />
    </>
  );
};

export default TabGroupComponent;
