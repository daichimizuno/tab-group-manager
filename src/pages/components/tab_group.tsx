import {
  Accordion,
  AccordionDetails,
  AccordionSummary,
  Typography,
} from "@mui/material";
import React from "react";
import { TabGroup } from "../../dao/chrome_storage_access";
import ExpandMore from "@mui/icons-material/ExpandMore";
import TabGroupTable from "./tab_group_table";

const TabGroupComponent = (props: any) => {
  const tabGroup = props.tabGroup as TabGroup[];

  return (
    <>
      <Typography data-testid="title" mb={4} mt={4}>
        登録されたタブグループ
      </Typography>

      <Accordion>
        <AccordionSummary
          expandIcon={<ExpandMore />}
          aria-controls="panel1a-content"
          id="panel1a-header"
        >
          <Typography>リスト1</Typography>
        </AccordionSummary>
        <AccordionDetails>
          <TabGroupTable tabGroup={tabGroup} />
        </AccordionDetails>
      </Accordion>
    </>
  );
};

export default TabGroupComponent;
