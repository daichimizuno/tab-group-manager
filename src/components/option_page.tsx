import React, { useEffect, useState } from "react";
import ReactDOM from "react-dom";
import ChromeStorageAccess, { Color, TabGroup } from "../chrome_storage_access";
import { TabGroupComponent } from "./tab_group";
const Options: React.FC = () => {
  const chromeStorage = new ChromeStorageAccess();

  // DB上のタブグループを保存するための変数
  const [tabGroup, setTabGroup] = useState<TabGroup[]>();

  // タブグループを作成する時の変数
  const [inputGroupName, setInputGroupName] = useState<string>("");
  const [selectedGroupColor, setSelectedTabGroupColor] =
    useState<string>("grey");

  // URLをグループに追加するための変数
  const [inputUrl, setInputUrl] = useState<string>("");
  const [selectedTabGroupNumber, setSelectedTabGroupNumber] =
    useState<number>(1);

  // DB上のTabGroupデータにView側のリストを依存させる
  useEffect(() => {
    getTabGroupData();
  }, [tabGroup]);

  // useEffectのためにDBとのTabGroupデータを動悸させる
  const syncTabGroup = () => {
    getTabGroupData();
  };

  const getTabGroupData = async () => {
    let returnTabGroup = await chromeStorage.getAllTabGroup();
    if (JSON.stringify(tabGroup) !== JSON.stringify(returnTabGroup)) {
      setTabGroup(returnTabGroup);
    }
  };

  const _createTabGroupData = async () => {
    if (inputGroupName.length > 0 && selectedGroupColor.length > 0) {
      setInputGroupName("");
      await chromeStorage.addNewTabGroup(
        inputGroupName,
        selectedGroupColor as Color
      );
    }

    syncTabGroup();
  };
  const _createUrlGroupData = async () => {
    await chromeStorage.addUrlToTabGroup(inputUrl, selectedTabGroupNumber);

    syncTabGroup();
  };

  const _haveTabGroup = (): boolean => {
    if (tabGroup !== undefined && tabGroup.length !== 0) {
      return true;
    } else {
      return false;
    }
  };

  return (
    <>
      <div className="new-tab-group">
        <p>新しいタブグループを作成</p>
        <input
          type="text"
          name="tabName"
          value={inputGroupName}
          placeholder="グループ名"
          onChange={(e) => setInputGroupName(e.target.value)}
        />
        <select onChange={(e) => setSelectedTabGroupColor(e.target.value)}>
          <option value="grey">grey</option>
          <option value="blue">blue</option>
          <option value="red">red</option>
          <option value="yellow">yellow</option>
          <option value="green">green</option>
          <option value="pink">pink</option>
          <option value="purple">purple</option>
          <option value="cyan">cyan</option>
        </select>
        <button
          data-testid="createTabGroupButton"
          onClick={_createTabGroupData}
        >
          作成
        </button>
      </div>

      <div className="new-url-to-group">
        <p>URLをグループに追加</p>
        <input
          style={{ width: "300px" }}
          type="text"
          name="url"
          placeholder="URL"
          onChange={(e) => {
            setInputUrl(e.target.value);
          }}
        />
        <select
          onChange={(e) => {
            setSelectedTabGroupNumber(Number(e.target.value));
          }}
        >
          {tabGroup?.map((tab) => {
            let tabId = tab.id;
            let tabName = tab.tabGroupName;
            return (
              <option value={tabId} key={tabId}>
                {tabName}
              </option>
            );
          })}
        </select>
        <button onClick={_createUrlGroupData}>作成</button>
      </div>

      {_haveTabGroup() ? (
        <TabGroupComponent tabGroup={tabGroup}> /</TabGroupComponent>
      ) : (
        <div></div>
      )}
    </>
  );
};

export default Options;
