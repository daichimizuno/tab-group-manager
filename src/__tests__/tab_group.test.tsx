import { render, screen } from "@testing-library/react";
import { tab } from "@testing-library/user-event/dist/tab";
import React from "react";
import renderer, { act, ReactTestRendererJSON } from "react-test-renderer";
import { TabGroup } from "../chrome_storage_access";
import { TabGroupComponent } from "../components/tab_group";

const tabGroupMock = [
  {
    id: 1,
    tabGroupName: "テスト",
    tabColor: "grey",
    urls: ["https://google.com"],
  },
  {
    id: 2,
    tabGroupName: "テスト2",
    tabColor: "red",
    urls: ["https://yahoo.co.jp"],
  },
] as TabGroup[];

let tree: ReactTestRendererJSON;

beforeEach(() => {
  tree = renderer.create(<TabGroupComponent tabGroup={tabGroupMock} />).toJSON() as ReactTestRendererJSON;
});

afterEach(() => {});

describe("tab_groupのテスト", () => {
  describe("コンポーネントのてすと", () => {
    test("スナップショットが前回の結果と一致しているかどうか", () => {
      expect(tree).toMatchSnapshot();
    });

    test("表示されるデータがTableがMockデータと正しいこと", () => {
      act(() => {
        render(<TabGroupComponent tabGroup={tabGroupMock} />);
      });

      const tabData1Node = screen.getAllByTestId(1) as any;
      const tabData2Node = screen.getAllByTestId(2) as any;

      expect(tabData1Node.length).toBe(4);
      expect(tabData2Node.length).toBe(4);

      expect(screen.getByText("1")).toBeTruthy();
      expect(screen.getByText("2")).toBeTruthy();
      expect(screen.getByText("テスト")).toBeTruthy();
      expect(screen.getByText("テスト2")).toBeTruthy();
      expect(screen.getByText("grey")).toBeTruthy();
      expect(screen.getByText("red")).toBeTruthy();
      expect(screen.getByText("https://google.com")).toBeTruthy();
      expect(screen.getByText("https://yahoo.co.jp")).toBeTruthy();
    });
  });
});
