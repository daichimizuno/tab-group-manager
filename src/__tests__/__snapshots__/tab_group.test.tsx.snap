// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tab_groupのテスト コンポーネントのてすと スナップショットが前回の結果と一致しているかどうか 1`] = `
<table>
  <thead
    className="table-dark"
  >
    <tr>
      <th
        scope="col-1"
      >
        id
      </th>
      <th
        scope="col-2"
      >
        色
      </th>
      <th
        scope="col-3"
      >
        タブ名
      </th>
      <th
        scope="col-4"
      >
        url
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td
        data-testid={1}
      >
        1
      </td>
      <td
        data-testid={1}
      >
        grey
      </td>
      <td
        data-testid={1}
      >
        テスト
      </td>
      <td
        data-testid={1}
      >
        https://google.com
      </td>
    </tr>
    <tr>
      <td
        data-testid={2}
      >
        2
      </td>
      <td
        data-testid={2}
      >
        red
      </td>
      <td
        data-testid={2}
      >
        テスト2
      </td>
      <td
        data-testid={2}
      >
        https://yahoo.co.jp
      </td>
    </tr>
  </tbody>
</table>
`;
